diff --git a/browser/base/jar.mn b/browser/base/jar.mn
index 824c5df20ee13af7af2ce8657506f6f3f457aafc..6ef49535676325768497c1804f91c9e93959199a 100644
--- a/browser/base/jar.mn
+++ b/browser/base/jar.mn
@@ -103,3 +103,5 @@ browser.jar:
 
 # L10n resources and overrides.
 % override chrome://global/locale/appstrings.properties chrome://browser/locale/appstrings.properties
+
+#include content/zen-assets.jar.inc.mn
