diff --git a/browser/components/tabbrowser/content/browser-ctrlTab.js b/browser/components/tabbrowser/content/browser-ctrlTab.js
index e5d16e605b7edf11fc9f52b93e415087b76398f1..feb7ccfc0562efda56177a75454a9f83a332e8d1 100644
--- a/browser/components/tabbrowser/content/browser-ctrlTab.js
+++ b/browser/components/tabbrowser/content/browser-ctrlTab.js
@@ -761,7 +761,7 @@ var ctrlTab = {
   _initRecentlyUsedTabs() {
     this._recentlyUsedTabs = Array.prototype.filter.call(
       gBrowser.tabs,
-      tab => !tab.closing && !tab.hidden
+      tab => !tab.closing && !tab.hidden && !tab.hasAttribute("zen-empty-tab")
     );
     this._sortRecentlyUsedTabs();
   },
