diff --git a/browser/base/content/browser.js b/browser/base/content/browser.js
index 019b168c1aeae7e1c97a3ae58c99a48a27f54134..56ed0bffb0ec0863c18ef881acb4efcc590e4a96 100644
--- a/browser/base/content/browser.js
+++ b/browser/base/content/browser.js
@@ -32,6 +32,7 @@ ChromeUtils.defineESModuleGetters(this, {
   ContextualIdentityService:
     "resource://gre/modules/ContextualIdentityService.sys.mjs",
   CustomizableUI: "resource:///modules/CustomizableUI.sys.mjs",
+  ZenCustomizableUI: "chrome://browser/content/ZenCustomizableUI.sys.mjs",
   DevToolsSocketStatus:
     "resource://devtools/shared/security/DevToolsSocketStatus.sys.mjs",
   DownloadUtils: "resource://gre/modules/DownloadUtils.sys.mjs",
@@ -3440,6 +3441,11 @@ var XULBrowserWindow = {
     AboutReaderParent.updateReaderButton(gBrowser.selectedBrowser);
     TranslationsParent.onLocationChange(gBrowser.selectedBrowser);
 
+    gZenViewSplitter.onLocationChange(gBrowser.selectedBrowser);
+    ZenWorkspaces.onLocationChange(gBrowser.selectedBrowser);
+    gZenTabUnloader.onLocationChange(gBrowser.selectedBrowser);
+    gZenPinnedTabManager.onLocationChange(gBrowser.selectedBrowser);
+
     PictureInPicture.updateUrlbarToggle(gBrowser.selectedBrowser);
 
     if (!gMultiProcessBrowser) {
@@ -4435,7 +4441,7 @@ nsBrowserAccess.prototype = {
     // Passing a null-URI to only create the content window,
     // and pass true for aSkipLoad to prevent loading of
     // about:blank
-    return this.getContentWindowOrOpenURIInFrame(
+    let res = this.getContentWindowOrOpenURIInFrame(
       null,
       aParams,
       aWhere,
@@ -4443,6 +4449,10 @@ nsBrowserAccess.prototype = {
       aName,
       true
     );
+    if (typeof window.gZenGlanceManager !== "undefined" && window.toolbar.visible) {
+      window.gZenGlanceManager.onTabOpen(res, aURI);
+    }
+    return res;
   },
 
   openURIInFrame: function browser_openURIInFrame(
@@ -6351,7 +6361,7 @@ function switchToTabHavingURI(
       ignoreQueryString || replaceQueryString,
       ignoreFragmentWhenComparing
     );
-    let browsers = aWindow.gBrowser.browsers;
+    let browsers = aWindow.ZenWorkspaces.allUsedBrowsers;
     for (let i = 0; i < browsers.length; i++) {
       let browser = browsers[i];
       let browserCompare = cleanURL(
@@ -6364,6 +6374,7 @@ function switchToTabHavingURI(
         continue;
       }
       if (requestedCompare == browserCompare) {
+        aWindow.ZenWorkspaces.switchIfNeeded(browser);
         // If adoptIntoActiveWindow is set, and this is a cross-window switch,
         // adopt the tab into the current window, after the active tab.
         let doAdopt =
