diff --git a/browser/base/content/browser.xhtml b/browser/base/content/browser.xhtml
index 68849f5093cd8ab99ab0747d972799eaf3ad7e22..3b4e6a179ab7b71859f832709778c326a39cffda 100644
--- a/browser/base/content/browser.xhtml
+++ b/browser/base/content/browser.xhtml
@@ -26,6 +26,7 @@
         sizemode="normal"
         retargetdocumentfocus="urlbar-input"
         scrolling="false"
+        zen-before-loaded="true"
         persist="screenX screenY width height sizemode"
         data-l10n-sync="true">
 <head>
@@ -100,6 +101,8 @@
 
   <title data-l10n-id="browser-main-window-default-title"></title>
 
+#include zen-preloaded.inc.xhtml
+
 # All JS files which are needed by browser.xhtml and other top level windows to
 # support MacOS specific features *must* go into the global-scripts.inc file so
 # that they can be shared with macWindow.inc.xhtml.
@@ -169,6 +172,7 @@
     });
   });
 </script>
+#include zen-assets.inc.xhtml
 </head>
 <html:body xmlns="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul">
 # All sets except for popupsets (commands, keys, and stringbundles)
@@ -190,9 +194,13 @@
     </vbox>
   </html:template>
 
-#include navigator-toolbox.inc.xhtml
-
-#include browser-box.inc.xhtml
+  <hbox id="zen-main-app-wrapper" flex="1" persist="zen-compact-mode">
+    <vbox id="zen-toast-container"></vbox>
+    #include navigator-toolbox.inc.xhtml
+    <html:span id="zen-sidebar-box-container">
+    </html:span>
+    #include browser-box.inc.xhtml
+  </hbox>
 
   <html:template id="customizationPanel">
     <box id="customization-container" flex="1" hidden="true"><![CDATA[
