diff --git a/image/decoders/nsJXLDecoder.h b/image/decoders/nsJXLDecoder.h
index 6cde7456ca03f79e74401c1d215b9d50453ebf41..2f593ca3b70100c600b86e753d7a458c83b4f15c 100644
--- a/image/decoders/nsJXLDecoder.h
+++ b/image/decoders/nsJXLDecoder.h
@@ -48,6 +48,18 @@ class nsJXLDecoder final : public Decoder {
   Vector<uint8_t> mBuffer;
   Vector<uint8_t> mOutBuffer;
   JxlBasicInfo mInfo{};
+  JxlPixelFormat mFormat;
+  JxlFrameHeader mFrameHeader;
+
+  bool mUsePipeTransform;
+  uint8_t mChannels;
+  uint8_t* mCMSLine;
+
+  uint32_t mNumFrames;
+  FrameTimeout mTimeout;
+  gfx::SurfaceFormat mSurfaceFormat;
+  SurfacePipe mPipe;
+  bool mContinue;
 };
 
 }  // namespace mozilla::image
