diff --git a/toolkit/components/pictureinpicture/PictureInPicture.sys.mjs b/toolkit/components/pictureinpicture/PictureInPicture.sys.mjs
index 5da0404b2672ba8cce7bcf808bf2373474776654..c3d58941b66c54f9d506698d015e294f8c8a5ceb 100644
--- a/toolkit/components/pictureinpicture/PictureInPicture.sys.mjs
+++ b/toolkit/components/pictureinpicture/PictureInPicture.sys.mjs
@@ -488,13 +488,13 @@ export var PictureInPicture = {
       return;
     }
 
-    let gBrowser = browser.getTabBrowser();
-    let tab = gBrowser.getTabForBrowser(browser);
+    let gBrowser = browser?.ownerGlobal?.gBrowser;
+    let tab = browser?.ownerGlobal?.gZenBrowserManagerSidebar.getTabForBrowser(browser) ?? gBrowser.getTabForBrowser(browser);
 
     // focus the tab's window
     tab.ownerGlobal.focus();
 
-    gBrowser.selectedTab = tab;
+    browser?.ownerGlobal?.ZenWorkspaces.switchIfNeeded(browser);
     await this.closeSinglePipWindow({ reason: "Unpip", actorRef: pipActor });
   },
 
@@ -877,7 +877,7 @@ export var PictureInPicture = {
     win.setIsMutedState(videoData.isMuted);
 
     // set attribute which shows pip icon in tab
-    let tab = parentWin.gBrowser.getTabForBrowser(browser);
+    let tab = parentWin.gBrowser.getTabForBrowser(browser) ?? parentWin.gZenBrowserManagerSidebar.getTabForBrowser(browser);;
     tab.setAttribute("pictureinpicture", true);
 
     this.setUrlbarPipIconActive(parentWin);
