diff --git a/Cargo.toml b/Cargo.toml
index 29942b65f62497b80d63675501ae063c393943a7..a9aa4641421038fbe5b0ecd9b7353611eecefffe 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -198,6 +198,9 @@ rure = { path = "third_party/rust/rure" }
 # Patch `plist` to work with `indexmap` 2.*
 plist = { path = "third_party/rust/plist" }
 
+# Patch mime_guess to add missing mime types
+mime_guess = { path = "third_party/rust/mime_guess" }
+
 # To-be-published changes.
 unicode-bidi = { git = "https://github.com/servo/unicode-bidi", rev = "ca612daf1c08c53abe07327cb3e6ef6e0a760f0c" }
 nss-gk-api = { git = "https://github.com/beurdouche/nss-gk-api", rev = "e48a946811ffd64abc78de3ee284957d8d1c0d63" }
